drivers: clock_control: add support for PLL3 on STM32 H7

gmarull opened the pull request:
Add support for enabling and configuring PLL3 on STM32 H7 series. PLL3
is used as a clock source by certain peripherals, e.g. LTDC.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
drivers: clock_control: provide function to compute PLL VCO input range

Provide a utility function to compute PLL VCO input range so that it
can be re-used for other PLLs.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

erwango commented on drivers/clock_control/clock_stm32_ll_h7.c:
```diff
@@ -298,7 +284,28 @@ static int32_t optimize_regulator_voltage_scale(uint32_t sysclk_freq)
 	defined(CONFIG_CLOCK_STM32_PLL_SRC_CSI)
 
-static int32_t get_vco_output_range(uint32_t vco_input_range)
+static int get_vco_input_range(uint32_t m_div, uint32_t *range)
+{
+	uint32_t vco_freq;
+
+	vco_freq = PLLSRC_FREQ / m_div;
+
+	if (1000000UL <= vco_freq && vco_freq <= 2000000UL) {
+		*range = LL_RCC_PLLINPUTRANGE_1_2;
+	} else if (2000000UL < vco_freq && vco_freq <= 4000000UL) {
+		*range = LL_RCC_PLLINPUTRANGE_2_4;
+	} else if (4000000UL < vco_freq && vco_freq <= 8000000UL) {
+		*range = LL_RCC_PLLINPUTRANGE_4_8;
+	} else if (8000000UL < vco_freq && vco_freq <= 16000000UL) {
+		*range = LL_RCC_PLLINPUTRANGE_8_16;
+	} else {
+		return -ERANGE;
+	}
+
+	return 0;
```
Can you move this part to a dedicated commit/PR ?

gmarull replied:
done

gmarull committed:
drivers: clock_control: add support for PLL3 on STM32 H7

Add support for enabling and configuring PLL3 on STM32 H7 series. PLL3
is used as a clock source by certain peripherals, e.g. LTDC.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

erwango commented on drivers/clock_control/clock_stm32_ll_h7.c:
```diff
@@ -684,6 +684,41 @@ static int stm32_clock_control_init(const struct device *dev)
+	LL_RCC_PLL3_SetVCOInputRange(vco_input_range);
+	LL_RCC_PLL3_SetVCOOutputRange(vco_output_range);
+
+#if defined(CONFIG_CLOCK_STM32_PLL3_P_ENABLE)
+	LL_RCC_PLL3P_Enable();
+	LL_RCC_PLL3_SetP(CONFIG_CLOCK_STM32_PLL3_P_DIVISOR);
+#endif /* CONFIG_CLOCK_STM32_PLL3_P_ENABLE */
+#if defined(CONFIG_CLOCK_STM32_PLL3_Q_ENABLE)
+	LL_RCC_PLL3Q_Enable();
+	LL_RCC_PLL3_SetQ(CONFIG_CLOCK_STM32_PLL3_Q_DIVISOR);
+#endif /* CONFIG_CLOCK_STM32_PLL3_Q_ENABLE */
+#if defined(CONFIG_CLOCK_STM32_PLL3_R_ENABLE)
+	LL_RCC_PLL3R_Enable();
+	LL_RCC_PLL3_SetR(CONFIG_CLOCK_STM32_PLL3_R_DIVISOR);
+#endif /* CONFIG_CLOCK_STM32_PLL3_R_ENABLE */
```
Nit: IS_ENABLED can be used here.

gmarull replied:
@erwango using `IS_ENABLED` leads to some compile errors because `*_DIVISOR` are conditionally defined, any suggestions?

erwango replied:
@gmarull Ha right, I was focusing on the LL_RCC_* functions. Nvm this is fine.

erwango commented:
@fzawadiak would you mind having a look and approve if that's fine with you ?

erwango commented on drivers/clock_control/clock_stm32_ll_h7.c:
```diff
@@ -298,7 +284,28 @@ static int32_t optimize_regulator_voltage_scale(uint32_t sysclk_freq)
 	defined(CONFIG_CLOCK_STM32_PLL_SRC_HSI) || \
 	defined(CONFIG_CLOCK_STM32_PLL_SRC_CSI)
```
```suggestion
	defined(CONFIG_CLOCK_STM32_PLL_SRC_CSI) || \
	defined(CONFIG_CLOCK_STM32_PLL3_ENABLE)
```
Apply everywhere

gmarull replied:
good catch, fixed

erwango replied:
@gmarull I am not seeing it fixed, did I missed something?

gmarull replied:
I missed one entry, sorry. can you check now?

Nukersson commented on drivers/clock_control/clock_stm32_ll_h7.c:
```diff
@@ -298,7 +284,28 @@ static int32_t optimize_regulator_voltage_scale(uint32_t sysclk_freq)
-static int32_t get_vco_output_range(uint32_t vco_input_range)
+static int get_vco_input_range(uint32_t m_div, uint32_t *range)
+{
+	uint32_t vco_freq;
+
+	vco_freq = PLLSRC_FREQ / m_div;
```
```suggestion
	const uint32_t vco_freq = PLLSRC_FREQ / m_div;
```

Not sure about `const`, maybe it need to be `static const`. Anyway the idea behind of this change - give compiler a chance to optimize better.

gmarull replied:
this is a stack variable, there's no need to do that

Nukersson commented on drivers/clock_control/clock_stm32_ll_h7.c:
```diff
@@ -298,7 +284,28 @@ static int32_t optimize_regulator_voltage_scale(uint32_t sysclk_freq)
+	uint32_t vco_freq;
+
+	vco_freq = PLLSRC_FREQ / m_div;
+
+	if (1000000UL <= vco_freq && vco_freq <= 2000000UL) {
+		*range = LL_RCC_PLLINPUTRANGE_1_2;
+	} else if (2000000UL < vco_freq && vco_freq <= 4000000UL) {
+		*range = LL_RCC_PLLINPUTRANGE_2_4;
+	} else if (4000000UL < vco_freq && vco_freq <= 8000000UL) {
+		*range = LL_RCC_PLLINPUTRANGE_4_8;
+	} else if (8000000UL < vco_freq && vco_freq <= 16000000UL) {
+		*range = LL_RCC_PLLINPUTRANGE_8_16;
+	} else {
+		return -ERANGE;
+	}
```
```suggestion

#define FREQ_MHz(x) (x*1000000UL)

	if (MHz(1) <= vco_freq && vco_freq <= MHz(2)) {
		*range = LL_RCC_PLLINPUTRANGE_1_2;
	} else if (MHz(2) < vco_freq && vco_freq <= MHz(4)) {
		*range = LL_RCC_PLLINPUTRANGE_2_4;
	} else if (MHz(4) < vco_freq && vco_freq <= MHz(8)) {
		*range = LL_RCC_PLLINPUTRANGE_4_8;
	} else if (MHz(8) < vco_freq && vco_freq <= MHz(16)) {
		*range = LL_RCC_PLLINPUTRANGE_8_16;
	} else {
		return -ERANGE;
	}
```

gmarull replied:
I'd prefer to discuss a generic macro such as this one at Zephyr level, not at module level

martinjaeger replied:
FYI:
https://github.com/zephyrproject-rtos/zephyr/blob/2d215484cc352152e9686477fead841c13ebb249/include/sys/util.h#L306

gmarull replied:
@martinjaeger great, was not aware of it!

gmarull replied:
fixed

Nukersson commented on drivers/clock_control/clock_stm32_ll_h7.c:
```diff
@@ -670,6 +684,41 @@ static int stm32_clock_control_init(const struct device *dev)
+#endif /* CONFIG_CLOCK_STM32_PLL3_R_ENABLE */
+
+	LL_RCC_PLL3_Enable();
+	while (LL_RCC_PLL3_IsReady() != 1) {
```
```suggestion
	while (LL_RCC_PLL3_IsReady() != 1U) {
```

gmarull replied:
fixed

KozhinovAlexander reviewed (commented):
Very small changes needed. But after that LGTM and good work!

KozhinovAlexander reviewed (approved):
LGTM
